File: sdk/ruby/lib/anyvali/schemas/record_schema.rb
# frozen_string_literal: true

module AnyVali
  class RecordSchema < Schema
    attr_reader :values_schema

    def initialize(values:, **kwargs)
      @values_schema = values
      super(kind: "record", **kwargs)
    end

    def to_node
      node = super
      node["values"] = @values_schema.to_node
      node
    end

    protected

    def validate(value, path, issues, context)
      unless value.is_a?(Hash)
        issues << ValidationIssue.new(
          code: IssueCodes::INVALID_TYPE,
          path: path,
          expected: "record",
          received: Schema.type_name(value)
        )
        return
      end

      value.each do |k, v|
        result = @values_schema.safe_parse(v, path: path + [k], context: context)
        issues.concat(result.issues) if result.failure?
      end
    end

    def dup_with(**overrides)
      values = overrides.delete(:values) || @values_schema
      attrs = {
        values: values,
        kind: @kind,
        constraints: @constraints,
        coerce_config: @coerce_config,
        default_value: @default_value,
        has_default: @has_default,
        custom_validators: @custom_validators
      }.merge(overrides)
      self.class.new(**attrs)
    end
  end
end

File: sdk/csharp/src/AnyVali/Schemas/RecordSchema.cs
namespace AnyVali.Schemas;

public sealed class RecordSchema : Schema<Dictionary<string, object?>>
{
    private readonly Schema _valueSchema;

    public RecordSchema(Schema valueSchema)
    {
        _valueSchema = valueSchema;
    }

    public new RecordSchema Default(object? value) => (RecordSchema)base.Default(value);

    internal override object? Validate(object? input, ValidationContext ctx)
    {
        if (input is not Dictionary<string, object?> obj)
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.InvalidType,
                Message = $"Expected record, received {DescribeType(input)}",
                Path = ctx.ClonePath(),
                Expected = "record",
                Received = DescribeType(input),
            });
            return null;
        }

        var result = new Dictionary<string, object?>();
        foreach (var (key, value) in obj)
        {
            ctx.PushPath(key);
            result[key] = _valueSchema.RunPipeline(value, ctx);
            ctx.PopPath();
        }
        return result;
    }

    internal override Dictionary<string, object?> ToNode()
    {
        var node = new Dictionary<string, object?>
        {
            ["kind"] = "record",
            ["valueSchema"] = _valueSchema.ToNode(),
        };
        AddDefaultAndCoercion(node);
        return node;
    }

    internal override Schema Clone() => new RecordSchema(_valueSchema)
    {
        DefaultValue = DefaultValue, CoercionCfg = CoercionCfg, IsPortable = IsPortable,
    };
}
